File: app/src/main/java/com/example/myapplication/LegendAdapter.java
package com.example.myapplication;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;
import java.util.Locale;

public class LegendAdapter extends RecyclerView.Adapter<LegendAdapter.VH> {

    private final List<LegendItem> items;
    private final double total;

    public LegendAdapter(List<LegendItem> items, double total) {
        this.items = items;
        this.total = total;
    }

    public static class VH extends RecyclerView.ViewHolder {
        public final View dot;
        public final TextView tvCategory;
        public final TextView tvAmount;
        public final TextView tvPercent;

        public VH(View view) {
            super(view);
            dot = view.findViewById(R.id.dot);
            tvCategory = view.findViewById(R.id.tvCategory);
            tvAmount = view.findViewById(R.id.tvAmount);
            tvPercent = view.findViewById(R.id.tvPercent);
        }
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_legend, parent, false);
        return new VH(view);
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        LegendItem item = items.get(position);
        holder.dot.setBackgroundColor(item.getColor());
        holder.tvCategory.setText(item.getLabel());
        holder.tvAmount.setText("₹" + (int) item.getAmount());
        double percent = (item.getAmount() / total) * 100;
        holder.tvPercent.setText(String.format(Locale.US, "%.1f%%", percent));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }
}

File: app/src/main/java/com/example/myapplication/LegendAdapter.kt
package com.example.myapplication

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

data class LegendItem(val label: String, val amount: Double, val color: Int)

class LegendAdapter(
    private val items: List<LegendItem>,
    private val total: Double
) : RecyclerView.Adapter<LegendAdapter.VH>() {

    class VH(view: View) : RecyclerView.ViewHolder(view) {
        val dot: View        = view.findViewById(R.id.dot)
        val tvCategory: TextView = view.findViewById(R.id.tvCategory)
        val tvAmount: TextView   = view.findViewById(R.id.tvAmount)
        val tvPercent: TextView  = view.findViewById(R.id.tvPercent)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        VH(LayoutInflater.from(parent.context).inflate(R.layout.item_legend, parent, false))

    override fun getItemCount() = items.size

    override fun onBindViewHolder(holder: VH, position: Int) {
        val item = items[position]
        holder.dot.setBackgroundColor(item.color)
        holder.tvCategory.text = item.label
        holder.tvAmount.text   = "₹${item.amount.toInt()}"
        holder.tvPercent.text  = "${"%.1f".format(item.amount / total * 100)}%"
    }
}
